# Data Cleaning

Cleaning and parsing of data from the [NBCUniversal Analytics Challenge](http://sc.aisnet.org/conference2018/student-competitions/nbcuniversal-challenge/).

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

## Loading Data

In [2]:
data = pd.read_csv('Data/NBCU-dataLaurel.csv')
data.head()

imdbid                        title  \
0  tt0010323  The Cabinet of Dr. Caligari   
1  tt0052893          Hiroshima Mon Amour   
2  tt0058898                   Alphaville   
3  tt0074252          Ugly, Dirty and Bad   
4  tt0084269                Losing Ground   

                                                plot     rating  imdb_rating  \
0  Hypnotist Dr. Caligari uses a somnambulist, Ce...    UNRATED          8.1   
1  A French actress filming an anti-war film in H...  NOT RATED          8.0   
2  A U.S. secret agent is sent to the distant spa...  NOT RATED          7.2   
3  Four generations of a family live crowded toge...        NaN          7.9   
4  A comedy-drama about a Black American female p...        NaN          6.3   

   metacritic dvd_release                          production  \
0         NaN   15-Oct-97                     Rialto Pictures   
1         NaN   24-Jun-03                     Rialto Pictures   
2         NaN   20-Oct-98                     Rialto Pictures   
3         NaN    1-Nov-16  Compagnia Cinematografica Champion   
4         NaN         NaN              Milestone Film & Video   

                                              actors imdb_votes  \
0  Werner Krauss, Conrad Veidt, Friedrich Feher, ...     42,583   
1  Emmanuelle Riva, Eiji Okada, Stella Dassas, Pi...     21,154   
2      Eddie Constantine, Anna Karina, Akim Tamiroff     17,801   
3  Nino Manfredi, Maria Luisa Santella, Francesco...      5,705   
4  Billie Allen, Gary Bolling, Clarence Branch Jr...        132   

                                              poster          director  \
0  https://images-na.ssl-images-amazon.com/images...      Robert Wiene   
1  https://images-na.ssl-images-amazon.com/images...     Alain Resnais   
2  https://images-na.ssl-images-amazon.com/images...   Jean-Luc Godard   
3  https://images-na.ssl-images-amazon.com/images...      Ettore Scola   
4  https://images-na.ssl-images-amazon.com/images...  Kathleen Collins   

  release_date  runtime                     genre  \
0    19-Mar-21   67 min  Fantasy, Horror, Mystery   
1    16-May-60   90 min            Drama, Romance   
2     5-May-65   99 min    Drama, Mystery, Sci-Fi   
3    23-Sep-76  115 min             Comedy, Drama   
4     1-Jun-82   86 min             Comedy, Drama   

                                              awards  \
0                                      1 nomination.   
1  Nominated for 1 Oscar. Another 6 wins & 5 nomi...   
2                                             1 win.   
3                             1 win & 2 nominations.   
4                                                NaN   

                                            keywords  Budget Box Office Gross  
0  expressionism|somnambulist|avant-garde|hypnosi...   18000                0  
1  memory|atomic-bomb|lovers-separation|impossibl...   88300                0  
2  dystopia|french-new-wave|satire|comic-violence...  220000            46585  
3  incest|failed-murder-attempt|poisoned-food|bap...    6590                0  
4  artist|painter|marriage|black-independent-film...       0                0

In [3]:
data.shape

(8468, 19)

These are the variables we have to work with:

imdbid: Unique Id used by IMDB to refer to the movie.

Title: Title of the movie

plot: Movie plot summary

rating: MPAA Appropriate audience rating

imdb_rating: IMDB's voters' scoring of a movie on a scale from 1-10 (10 being best)

metacritic: Metacritic movie score on a scale of 0-100 (100 being best)

dvd_release: Movie release date on DVD

production: Principle production company

actors: Lead Actors

imdb_votes: Total votes from IMDB members

poster: Movie Poster artwork

director: Movie director

release_date: Theatrical Release Date

runtime: Runtime length of movie in minutes

genre: Genre Classification

awards: Academy awards & nominations

keywords: Keywords associated with the movie

budget: Budget spent on movie production, marketing, and distribution

box office gross: Box Office Gross Returns as of 9/21/2017

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8468 entries, 0 to 8467
Data columns (total 19 columns):
imdbid              8468 non-null object
title               8468 non-null object
plot                8196 non-null object
rating              5252 non-null object
imdb_rating         7735 non-null float64
metacritic          5079 non-null float64
dvd_release         5335 non-null object
production          6758 non-null object
actors              8153 non-null object
imdb_votes          7735 non-null object
poster              7967 non-null object
director            8390 non-null object
release_date        8283 non-null object
runtime             7846 non-null object
genre               8424 non-null object
awards              5242 non-null object
keywords            6381 non-null object
Budget              8468 non-null object
Box Office Gross    8468 non-null object
dtypes: float64(2), object(17)
memory usage: 1.2+ MB


Notice how many of variables are just objects. We're going to have to deal with converting a few of these into useful types.

### Parsing imdbid

The proper format for imdb id's is simply 7 digits, with no tt.

In [5]:
data['imdbid'] = data['imdbid'].str.replace('tt','')

### Parsing release_dates and dvd_release

Next, we'll change release_date to a datetime-like type.

In [6]:
data['release_date'].head()

0    19-Mar-21
1    16-May-60
2     5-May-65
3    23-Sep-76
4     1-Jun-82
Name: release_date, dtype: object

In [7]:
pd.to_datetime(data['release_date'])[1],data['release_date'][1]

(Timestamp('2060-05-16 00:00:00'), '16-May-60')

Then we see that there is an issue with pandas to_datetime function. It converts very old dates back to the 19th century. Perhaps we need to use the datetime package per [this](https://stackoverflow.com/questions/16600548/how-to-parse-string-dates-with-2-digit-year).

In [8]:
dates = data['release_date']
dates = pd.to_datetime(dates)
opivot_index = dates[dates.apply(lambda x: x.year>2019)].index
for index in opivot_index:
    dates[index] = dates[index].replace(year = dates[index].year-100)
sum(dates>'2019-12-31')

0

Then we've found a way to account for Python's default pivot year.

In [9]:
data['release_date']=dates
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8468 entries, 0 to 8467
Data columns (total 19 columns):
imdbid              8468 non-null object
title               8468 non-null object
plot                8196 non-null object
rating              5252 non-null object
imdb_rating         7735 non-null float64
metacritic          5079 non-null float64
dvd_release         5335 non-null object
production          6758 non-null object
actors              8153 non-null object
imdb_votes          7735 non-null object
poster              7967 non-null object
director            8390 non-null object
release_date        8283 non-null datetime64[ns]
runtime             7846 non-null object
genre               8424 non-null object
awards              5242 non-null object
keywords            6381 non-null object
Budget              8468 non-null object
Box Office Gross    8468 non-null object
dtypes: datetime64[ns](1), float64(2), object(16)
memory usage: 1.2+ MB


It seems natural to also do the same for dvd_release.

In [10]:
dates = data['dvd_release']
dates = pd.to_datetime(dates)
opivot_index = dates[dates.apply(lambda x: x.year>2019)].index
for index in opivot_index:
    dates[index] = dates[index].replace(year = dates[index].year-100)
data['dvd_release'] = dates
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8468 entries, 0 to 8467
Data columns (total 19 columns):
imdbid              8468 non-null object
title               8468 non-null object
plot                8196 non-null object
rating              5252 non-null object
imdb_rating         7735 non-null float64
metacritic          5079 non-null float64
dvd_release         5335 non-null datetime64[ns]
production          6758 non-null object
actors              8153 non-null object
imdb_votes          7735 non-null object
poster              7967 non-null object
director            8390 non-null object
release_date        8283 non-null datetime64[ns]
runtime             7846 non-null object
genre               8424 non-null object
awards              5242 non-null object
keywords            6381 non-null object
Budget              8468 non-null object
Box Office Gross    8468 non-null object
dtypes: datetime64[ns](2), float64(2), object(15)
memory usage: 1.2+ MB


### Parsing imdb_votes

Next, we have several important numerical variables that are currenly in object types. First, we'll work with imdb_votes.

In [11]:
data['imdb_votes'].head()

0    42,583
1    21,154
2    17,801
3     5,705
4       132
Name: imdb_votes, dtype: object

So we need to convert imdb_votes to integers. Since there are commas in each number, we cannot simply tell pandas to treat each entry as an integer via the .astype() function. We'll first have to replace each comma with a blank, then apply the int() function. We also have to take care to ignore all of the missing values from imdb_votes as we will be dealing with those later.

In [12]:
votes = data['imdb_votes']
votes_parse = votes.str.replace(',','')
votes_parse.head()

0    42583
1    21154
2    17801
3     5705
4      132
Name: imdb_votes, dtype: object

Then, we've successfully removed all of the commas. Let's confirm that we didn't lose any datapoints along the way.

In [13]:
[len(votes_parse), len(data['imdb_votes'])]

[8468, 8468]

In order to convert to int, we have to find a way to work around missing values. Let's replace all the missing values with -1 and then convert them back to NaN after conversion.

In [14]:
import numpy as np

votes_int = votes_parse.fillna(-1).astype('int')
votes_int[votes_int==-1] = np.nan
votes_int.isna().sum() == votes_parse.isna().sum()

True

In [15]:
data['imdb_votes'] = votes_int
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8468 entries, 0 to 8467
Data columns (total 19 columns):
imdbid              8468 non-null object
title               8468 non-null object
plot                8196 non-null object
rating              5252 non-null object
imdb_rating         7735 non-null float64
metacritic          5079 non-null float64
dvd_release         5335 non-null datetime64[ns]
production          6758 non-null object
actors              8153 non-null object
imdb_votes          7735 non-null float64
poster              7967 non-null object
director            8390 non-null object
release_date        8283 non-null datetime64[ns]
runtime             7846 non-null object
genre               8424 non-null object
awards              5242 non-null object
keywords            6381 non-null object
Budget              8468 non-null object
Box Office Gross    8468 non-null object
dtypes: datetime64[ns](2), float64(3), object(14)
memory usage: 1.2+ MB


### Parsing Budget and Box Office Gross

Now we have to deal with 'Budget' and 'Box Office Gross' in a similar manner.

In [16]:
data[['Budget', 'Box Office Gross']].head()

Budget Box Office Gross
0   18000                0
1   88300                0
2  220000            46585
3    6590                0
4       0                0

Looks like we don't have to worry about any commas in 'Budget' or in 'Box Office Gross', so the conversions will be much simplier. In the previous revision of Data Cleaning, we found that data entries containing Box Office Gross values in foreign currencies were not worth saving. We will be removing them in this revision.

In [17]:
data = data.drop(data[data['Box Office Gross'].str.contains('GBP')].index)
data = data.drop(data[data['Box Office Gross'].str.contains('EU')].index)
data = data.drop(data[data['Budget'].str.contains('EU')].index)
data = data.drop(data[data['Budget'].str.contains('CAD')].index)

Now that we've removed all datapoints with foreign currencies, we can start parsing Budget and Box Office Gross into numerical types.

In [18]:
gross = data['Box Office Gross']
gross = gross.str.replace(',','')

error_index = []

for i, item in enumerate(gross):
    try:
        int(item)
    except ValueError:
        error_index.append(i)

In [19]:
gross[error_index]

/usr/local/lib/python3.7/site-packages/pandas/core/series.py:942: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  return self.loc[key]


1394                                                    0
1869                                                    0
2914                                                    0
3866                                                    0
5093                                             14793904
5684                                                    0
5723                                                    0
5828                                                18202
5981    <strongitemprop="name">AGrosslyunderratedmovie...
6277                                                  NaN
6902                                                    0
7075                                                    0
7137                                                    0
Name: Box Office Gross, dtype: object

This is peculiar, the value of Box Office Gross seems to change once we store that variable as gross (from these string entries to the number 0). On top of that, the original value in the dataset seems to be random text, perhaps a misscraped comment?

**Edit**: On review, enumerate gives us an enumerated index, so to find the erroneous entries, we need to use df.iloc

When I review the data entries for 1394, it's hard to see what relevance '<strongitemprop="name">Grossbuttotallyworthwhile</strong>' has. Then, we'll remove these samples from the dataset.

In [20]:
error_index = data.iloc[error_index].index
data = data.drop(error_index)

Then, we can finally change box office gross and budget into integer formats.

In [21]:
data['Box Office Gross'] = data['Box Office Gross'].astype('int')
data['Budget'] = data['Budget'].astype('int')
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8206 entries, 0 to 8467
Data columns (total 19 columns):
imdbid              8206 non-null object
title               8206 non-null object
plot                7951 non-null object
rating              5092 non-null object
imdb_rating         7510 non-null float64
metacritic          4904 non-null float64
dvd_release         5172 non-null datetime64[ns]
production          6568 non-null object
actors              7904 non-null object
imdb_votes          7510 non-null float64
poster              7729 non-null object
director            8131 non-null object
release_date        8028 non-null datetime64[ns]
runtime             7607 non-null object
genre               8163 non-null object
awards              5065 non-null object
keywords            6177 non-null object
Budget              8206 non-null int64
Box Office Gross    8206 non-null int64
dtypes: datetime64[ns](2), float64(3), int64(2), object(12)
memory usage: 1.3+ MB


### Parsing imdb_rating

Since imdb_rating has many unique values, we're going to round all the ratings down to the whole number. We can use train on the ranks themselves in decision trees, but we'll have to resort to one-hot encoding for regression, neural nets, etc.

In [22]:
data['imdb_rating'].isna().sum()

696

There are 696 data points missing imdb ratings. We'll try using imdb packages like [imdbpy](https://imdbpy.readthedocs.io/en/latest/) to fill in these values.

**Edit**: Turns out, imdbpy is only useful for getting the release *year* of a movie, not the release *date*. For now, it seems like all we can do is remove these data points.

In [23]:
na_index = data[data['imdb_rating'].isna()].index
data = data.drop(na_index)

In [24]:
import math

data['imdb_rating'] = data['imdb_rating'].apply(lambda x: math.floor(x))

### Parsing runtime

For runtime, we would like to remove the 'min' part of each entry. It's much better for analysis to just assume that runtime is measured in minutes.

In [25]:
data['runtime'].isna().sum()

142

In [26]:
from imdb import IMDb
ia = IMDb()

Unlike previously, imdbpy actualy has a feature to get movie runtimes from imdb. Because of this, we will be able to fill in the missing values.

In [27]:
na_index = data[data['runtime'].isna()].index
for index in na_index:
    movie = ia.get_movie(data.loc[index]['imdbid'])
    try:
        data.loc[index]['imdbid'] = movie['runtime']
    except KeyError:
        data = data.drop(index)

/usr/local/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [28]:
data[data['runtime'].isna()]

imdbid                                              title  \
832   1935277                                     Road to Juarez   
1989  3305388                                       Mountain Top   
2267  3640942                                       Koyelaanchal   
2428  3822606                                   Ra Ra Krishnayya   
2530  3969208                                         Trust Fund   
2588  4074296                              From This Day Forward   
2840  4489160                                     You're My Boss   
2900  4621100                                   Nanak Shah Fakir   
2914  4641602                                           365 Days   
3069  4944460                                Everyday I Love You   
3114  5066818                                  Remember the Goal   
3174  5176486                                    Grandma's House   
3183  5195438                     Spirit Game: Pride of a Nation   
3293  5454690                                      Love Is Blind   
3384  5669936                                         The Recall   
3529  6130166                                                Ism   
3532  6150238                                          The Mayor   
3564  6323258                         Naanna Nenu Naa Boyfriends   
3618  6750852                                      Rabb Da Radio   
3624  6789386                                              Posto   
4650  0495024                                          Bad Blood   
5043  1002963                              Y.M.I. Yeh Mera India   
5273  1130091                                        Solar Flare   
5278  1132606                                     Ugly Aur Pagli   
5623  1288644                          Mere Khwabon Mein Jo Aaye   
5624  1289419                                           N-Secure   
5937  1492705               A Little Something for Your Birthday   
6076  1573482                                            Striker   
6087  1582478                                            Bindaas   
6088  1582481                                        Blade Babji   
...       ...                                                ...   
7018  2262212                                    Taur Mittran Di   
7063  2304835                                       Mr. Nookayya   
7076  2316756                                    Andala Rakshasi   
7217  2531334                                      Keep Watching   
7246  2633826                                      Mumbai Mirror   
7290  2777850                                       Guru En Aalu   
7360  2996804                              Girl boy bakla tomboy   
7362  3004572                                               2101   
7435  3253548                        Naveena Saraswathi Sabatham   
7635  3960994                                            Shutter   
7674  4125206                 Lord of the Dance: Dangerous Games   
7701  4209788                                       Molly's Game   
7760  4472678                                         3600 Detik   
7807  4663708                                         360 Degree   
7818  4697186                               The Space Between Us   
7822  4706602                              Mega Mindy Versus ROX   
7943  5109746                  The Wrecking Crew vs. The Zombies   
7961  5143890                                            Justice   
7968  5175450  Mark Felt: The Man Who Brought Down the White ...   
7978  5204766                              Must Date the Playboy   
7981  5230196                                            Carface   
7997  5295670                          Mama Manchu Alludu Kanchu   
8015  5377658                          Dylan Moran: Off the Hook   
8022  5451690                                     American Satan   
8047  5540622                                      Alexander IRL   
8292  6490532                                       Buried Child   
8310  6604334                     Bob the Builder: Mega Machine

In [29]:
na_index = data[data['runtime'].isna()].index
runtimes = []
for index in na_index:
    movie = ia.get_movie(data.loc[index]['imdbid'])
    try:
        runtimes.append(movie['runtime'][0])
    except KeyError:
        data = data.drop(index)

In [30]:
data['runtime'].isna().sum()

72

In [31]:
na_data = data[data['runtime'].isna()]

In [32]:
for index in na_data.head().index:
    movie = ia.get_movie(na_data.loc[index]['imdbid'])
    na_data.loc[index]['runtime'] = movie['runtime']

/usr/local/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [33]:
na_data.head().index

Int64Index([832, 1989, 2267, 2428, 2530], dtype='int64')

In [34]:
movie = ia.get_movie('1935277')
na_data.loc[832]['runtime'] = movie['runtime']

/usr/local/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [35]:
runtimes1 = na_data['imdbid'].apply(lambda x: ia.get_movie(x)['runtime'][0])

In [36]:
na_data['runtime'] = na_data['runtime'].fillna(runtimes1)

/usr/local/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [37]:
data['runtime'] = data['runtime'].fillna(runtimes1)

In [38]:
data['runtime'].isna().sum()

0

Now that we've filled in all of the missing runtimes that we could possibly fill in, we can drop ' min' and convert to integer type.

In [39]:
data['runtime'] = data['runtime'].str.replace(' min', '')
data['runtime'] = data['runtime'].astype('int')

### Parsing genre

Now let's parse through genre in order to use it as a categorical variable.

In [40]:
genres = data['genre']

Thankfully, imdbpy has a function to retrieve genres for movies when available. We're going to attempt to fill in these values and drop the ones in which we can't.

In [41]:
na_genres = genres[genres.isna()].index

In [42]:
genres = []
for index in na_genres:
    movie = ia.get_movie(data.loc[index]['imdbid'])
    try:
        genres.append(movie['genres'][0])
    except KeyError:
        data = data.drop(index)

In [43]:
na_data = data[data['genre'].isna()]
genres1 = na_data['imdbid'].apply(lambda x: ia.get_movie(x)['genre'][0])

In [44]:
data['genre'] = data['genre'].fillna(genres1)

In [45]:
genre_index = data['genre'].index

In [46]:
for i in genre_index:
    parsed_genre = [x.strip() for x in data.loc[i]['genre'].split(',')]

In [47]:
genres = data['genre']

In [48]:
parsed_genres = genres.apply(lambda x: [y.strip() for y in x.split(',')])

In [49]:
genres = pd.Series(parsed_genres.sum()).unique()

Now that we've separated all of the distinct genres out, we have to create categorical variables for each.

In [50]:
for genre in genres:
    print(genre)

Fantasy
Horror
Mystery
Drama
Romance
Sci-Fi
Comedy
Crime
Action
Adventure
Animation
Family
History
Thriller
Biography
Documentary
Music
Sport
Western
Musical
War
News
Short


In [81]:
for i in range(len(parsed_genres[0])):
    print(parsed_genres[0][i])

Fantasy
Horror
Mystery


In [85]:
test = pd.DataFrame([genres])

In [91]:
data.shape[0]

7439

Now that we've separated out all of the genres in 'genre', we can create categorical variables for each category. Then, we can apply one-hot encoding to encode movie genres as categorical variables.

In [95]:
genre_classes = pd.DataFrame(0, index = data.index, columns = genres)

In [100]:
parsed_genres[0][i]

'Mystery'

In [108]:
index = 8464

In [112]:
genre_classes.loc[index][parsed_genres[index]] = 1

In [115]:
for index in data.index:
    genre_classes.loc[index][parsed_genres[index]] = 1

In [133]:
test = pd.merge(data, genre_classes, left_index=True, right_index=True)

In [140]:
test[genres].astype('category')

Fantasy Horror Mystery Drama Romance Sci-Fi Comedy Crime Action  \
0          1      1       1     0       0      0      0     0      0   
1          0      0       0     1       1      0      0     0      0   
2          0      0       1     1       0      1      0     0      0   
3          0      0       0     1       0      0      1     0      0   
4          0      0       0     1       0      0      1     0      0   
5          0      0       0     1       0      0      0     1      0   
6          0      0       0     0       0      0      1     0      1   
7          1      0       0     0       0      1      0     0      0   
8          1      0       0     0       0      0      0     0      0   
9          0      0       0     1       0      0      0     1      0   
10         0      0       0     1       1      0      0     0      0   
11         0      0       0     0       0      1      0     0      0   
12         0      0       0     1       1      0      1     0      0   
13         0      0       0     1       0      0      0     0      0   
14         0      0       0     1       1      0      0     0      0   
15         1      0       0     0       0      0      0     0      1   
16         0      0       0     0       0      0      0     1      1   
17         0      0       0     1       0      0      0     1      1   
18         0      0       0     0       0      0      1     0      0   
19         0      0       0     1       0      0      1     0      0   
20         0      0       1     1       0      0      0     1      0   
21         0      0       0     0       0      1      0     0      1   
22         0      0       0     1       0      0      1     0      0   
23         0      0       0     1       0      0      0     1      0   
24         0      0       0     0       0      0      1     0      0   
25         1      0       0     1       0      0      0     0      0   
26         0      0       0     0       0      0      1     0      0   
27         0      0       0     0       0      0      1     0      0   
28         0      0       0     0       0      1      0     0      1   
29         0      0       0     1       1      0      1     0      0   
...      ...    ...     ...   ...     ...    ...    ...   ...    ...   
8434       0      0       1     1       0      0      0     0      0   
8435       0      0       0     1       0      0      0     0      0   
8436       0      0       0     0       0      0      0     0      0   
8437       0      0       0     1       1      0      1     0      0   
8438       0      0       0     1       1      0      1     0      0   
8439       0      0       0     1       0      0      1     0      0   
8441       0      0       0     1       0      0      0     1      1   
8442       0      0       0     0       0      0      0     0      0   
8443       0      0       0     1       0      0      0     1      1   
8444       0      0       0     1       0      0      1     0      0   
8445       0      0       0     1       1      0      0     0      0   
8446       0      0       0     1       0      0      0     0      0   
8448       0      1       0     0       0      0      0     0      0   
8449       0      0       0     0       0      0      0     0      0   
8450       0      0       0     0       0      0      0     0      0   
8451       0      0       0     1       0      0      0     1      0   
8452       0      0       0     1       0      0      0     1      0   
8453       0      0       0     1       0      0      1     0      0   
8454       0      0       0     1       0      0      0     0      0   
8455       0      0       0     0       0      0      0     0      0   
8456       0      0       0     1       0      0      0     0      0   
8457       0      0       0     0       1      0      1     0      0   
8458       0      0       0     0       0      0      0     0      1   
8459       0      0       0     1       0    